fuzzbench/worker.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Self-defined worker module."""
+import time
+
+import redis
+import rq
+
+MAX_TIME_LIMIT = 3600
+REASSIGN_GAP_TIME = 5
+
+
+def no_pending_jobs(queue):
+    """Checks whether the queue has unfinished jobs."""
+    return (queue.deferred_job_registry.count +
+            queue.scheduled_job_registry.count +
+            queue.started_job_registry.count) == 0
+
+
+def main():
+    """Sets up Redis connection and starts the worker."""
+    redis_connection = redis.Redis(host="queue-server")
+    with rq.Connection(redis_connection):
+        queue = rq.Queue('build_n_run_queue')
+        worker = rq.Worker([queue], connection=redis_connection)
+
+        start_time = time.time()
+        while time.time() - start_time < MAX_TIME_LIMIT:
+            if not no_pending_jobs(queue):
+                worker.work(burst=True)
+            else:
+                time.sleep(REASSIGN_GAP_TIME)
+                if no_pending_jobs(queue):
+                    return
+            time.sleep(REASSIGN_GAP_TIME)
```

Comment:
would something like this work?
```
while len(queue) or len(queue.deferred_job_registry):
  worker.work(burst=True)
  time.sleep(5)
```

Reply:
This is interesting. I will try this later.
It should work if there is no multi-dependency. But since the current implementation has to resolve multi-denpendency, it probably will be broken when all old jobs finished but new jobs have not been scheduled.

Reply:
If we eliminate it as this plan:
```
Eliminate base-runner and the double-dependency of the benchmark-runner image
Eliminate base-runner
by moving the steps of base-runner into benchmark-runner ({fuzzer}-{benchmark}-runner)
This would allow builds of benchmark-runner to only depend on runner.Dockerfile builds which will depend on builder.Dockerfile but inherit from base-image. This is fine because builder.Dockerfile inherits/depends on base-image because 
Make {fuzzer}-{benchmark}-intermediate-runner depend on {fuzzer}-{benchmark}-builder
Make {fuzzer}-{benchmark}-runner only depend on {fuzzer}-{benchmark}-intermediate-runner 
```
Then it works for your proposed way.
```
while len(queue) or len(queue.deferred_job_registry):
  worker.work(burst=True)
  time.sleep(5)
```
Check this PR: https://github.com/zchcai/fuzzbench/pull/10.

Reply:
Thanks! Should we check-in the elimination of the double dependency first, and then this PR?

Reply:
Sounds good to me. I think the possible order would be: #584, then eliminate double dependency (which may have conflicts with #584), then #585 (this PR but use the way in zchcai#10).

Reply:
SG!